R/qenv-class.R

```diff
@@ -14,11 +14,60 @@
 #' @exportClass qenv
 setClass(
   "qenv",
-  slots = c(env = "environment", code = "character", id = "integer", warnings = "character", messages = "character"),
-  prototype = list(
-    env = new.env(parent = parent.env(.GlobalEnv)), code = character(0), id = integer(0),
-    warnings = character(0), messages = character(0)
-  )
+  slots = c(
+    code = "character",
+    id = "integer",
+    warnings = "character",
+    messages = "character"
+  ),
+  contains = "environment"
+)
+
+#' It initializes the `qenv` class
+#' @noRd
+setMethod(
+  "initialize",
+  "qenv",
+  function(.Object, # nolint: object_name.
+           .xData, # nolint: object_name.
+           code = character(0L),
+           warnings = rep("", length(code)),
+           messages = rep("", length(code)),
+           id = integer(0L),
+           ...) {
+    # # Pre-process parameters to ensure they are ready to be used by parent constructors
+    stopifnot("`code` must be a character or language object." = any(is.language(code), is.character(code)))
+
+    if (is.language(code)) {
+      code <- paste(lang2calls(code), collapse = "\n")
+    }
+    if (length(code)) {
+      code <- paste(code, collapse = "\n")
+    }
+
+    if (length(id) == 0L) {
+      id <- sample.int(.Machine$integer.max, size = length(code))
+    }
+
+    new_xdata <- if (rlang::is_missing(.xData)) {
+      new.env(parent = parent.env(.GlobalEnv))
+    } else {
+      checkmate::assert_environment(.xData)
+      rlang::env_clone(.xData, parent = parent.env(.GlobalEnv))
+    }
+    lockEnvironment(new_xdata, bindings = TRUE)
+
+    # .xData needs to be unnamed as the `.environment` constructor allows at
+    # most 1 unnamed formal argument of class `environment`.
+    # See methods::findMethods("initialize")$.environment
+    .Object <- methods::callNextMethod( # nolint: object_name.
+      # Mandatory use of `xData` to build a correct .Object@.xData
+      .Object, new_xdata,
+      code = code, messages = messages, warnings = warnings, id = id, ...
+    )
```

Comment:
This call is essential with `.xData` argument being mandatory!

Without it, the `@.xData` is the same between different instances on `qenv` (scary stuff and why I was delayed today to push this)

``` r
setClass(
  "an_env",
  contains = "environment"
)

setMethod(
  "initialize",
  signature = "an_env",
  definition = function(.Object, ...) {
    .Object
  }
)
q <- methods::new("an_env")
q@.xData$a <- 2
as.list(q)
#> $a
#> [1] 2

q2 <- methods::new("an_env")
as.list(q2)
#> $a
#> [1] 2
```

<sup>Created on 2024-11-06 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Reply:
This is interesting finding. I wonder how it works exactly as environment is not defined anywhere by us in the package. It is created inside of the function. Let's have a quick call after daily if you can

Reply:
Let's!

Reply:
Okey, I see now how it works.